### PR TITLE
DR2-1997 Create methods for flow control lambda

### PR DESCRIPTION
**Pull request by MancunianSam**
There are two added to the SFNClient

* List step function executions
* Send task success

There is a new SSMClient with one method

* Get parameter

I've also refactored the SFNClient tests to get rid of Mockito.

### CODE REVIEW

File: sfn/src/test/scala/uk/gov/nationalarchives/DASFNClientTest.scala [hunk 1]
```diff
@@ -3,66 +3,183 @@ package uk.gov.nationalarchives
...
+    )
+    val sfnAsyncClient = createClient(initialExecutions)
+    val client = DASFNClient[IO](sfnAsyncClient)
+
+    client.sendTaskSuccess("taskToken").unsafeRunSync()
+  }
+
+  "sendTaskSuccess" should "return an error if the task token is invalid" in {
+    val initialExecutions = ListBuffer(
+      SfnExecutions("name1running", "arn1", "", "running", "taskToken")
```
Comment (techncl) [inline sfn/src/test/scala/uk/gov/nationalarchives/DASFNClientTest.scala:164]: 1. Maybe `taskToken` could be `invalidTaskToken`
2. Maybe line 170 could be `client.sendTaskSuccess( initialExecutions.taskToken).unsafeRunSync()`?

File: ssm/src/test/scala/uk/gov/nationalarchives/DASSMClientTest.scala [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+    val ex = intercept[Exception] {
+      client.getParameter[Result]("testName").unsafeRunSync()
+    }
+
+    ex.getMessage should equal("Error decoding encrypted parameter")
+  }
+
+  "DASSMClient" should "return an error if the parameter is not found" in {
+    val ssmAsyncClient = createClient(ListBuffer(TestParameter("testName", true, """{"value": "testValue"}""")))
```
Comment (techncl) [inline ssm/src/test/scala/uk/gov/nationalarchives/DASSMClientTest.scala:70]: 1. Should `"testName"` be `"invalidTestName"`? 
2. Maybe line 75 could be`client.getParameter[Result](ssmAsyncClient.name).unsafeRunSync()`?

File: sfn/src/main/scala/uk/gov/nationalarchives/DASFNClient.scala [hunk 3]
```diff
@@ -34,9 +44,39 @@ trait DASFNClient[F[_]: Async]:
...
+  def listStepFunctions(stepFunctionArn: String, status: Status): F[List[String]]
+
+  def sendTaskSuccess(taskToken: String): F[Unit]
+
 object DASFNClient:
 
+  extension [F[_]: Async, T](completableFuture: CompletableFuture[T])
+    private def liftF: F[T] = Async[F].fromCompletableFuture(Async[F].pure(completableFuture))
+
+  enum Status(val sdkStatus: String):
```
Comment (techncl) [inline sfn/src/main/scala/uk/gov/nationalarchives/DASFNClient.scala:56]: Should we add all of [the statuses](https://docs.aws.amazon.com/step-functions/latest/apireference/API_DescribeExecution.html#StepFunctions-DescribeExecution-response-status)?

**Review by techncl — COMMENTED** [review]
Looks good! Just some minor comments.

File: ssm/src/main/scala/uk/gov/nationalarchives/DASSMClient.scala [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+   * @param withDecryption
+   *   Whether to decrypt the parameter
+   * @tparam T
+   *   The return type of the decoded string
+   * @return
+   *   An object T decoded from the parameter value
+   */
+  def getParameter[T](parameterName: String, withDecryption: Boolean = false)(using Decoder[T]): F[T]
+
+object DASSMClient:
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply ssm/src/main/scala/uk/gov/nationalarchives/DASSMClient.scala:34]: Yeah OK.

File: ssm/src/test/scala/uk/gov/nationalarchives/DASSMClientTest.scala [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+import software.amazon.awssdk.services.ssm.model.{GetParameterRequest, GetParameterResponse, Parameter, ParameterNotFoundException}
+
+import java.util.concurrent.CompletableFuture
+import scala.collection.mutable.ListBuffer
+
+class DASSMClientTest extends AnyFlatSpec:
+  case class Result(value: String)
+
+  case class TestParameter(name: String, encrypted: Boolean, value: String)
+  case class Errors(getParameter: Boolean = false)
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply ssm/src/test/scala/uk/gov/nationalarchives/DASSMClientTest.scala:18]: It matches the method name it causes the error for so I'll leave it as is.

File: sfn/src/test/scala/uk/gov/nationalarchives/DASFNClientTest.scala (hunk 1 above)
Comment (MancunianSam) [reply sfn/src/test/scala/uk/gov/nationalarchives/DASFNClientTest.scala:164]: This is testing that when we send a task token that's incorrect, we get an exception. So the task token on line 164 is a valid running task token and the one on line 170 is the invalid one which causes the exception. They shouldn't be the same.

File: ssm/src/test/scala/uk/gov/nationalarchives/DASSMClientTest.scala (hunk 2 above)
Comment (MancunianSam) [reply ssm/src/test/scala/uk/gov/nationalarchives/DASSMClientTest.scala:70]: Similar comment to the previous one, line 70 is the valid initial state and then we're testing what happens if we pass in invalid state.

File: sfn/src/main/scala/uk/gov/nationalarchives/DASFNClient.scala (hunk 3 above)
Comment (MancunianSam) [reply sfn/src/main/scala/uk/gov/nationalarchives/DASFNClient.scala:56]: Yeah we should, good idea. If we get one in one of the missing ones it will fail.

**Review by techncl — APPROVED** [review]
👍

**Pull request merged**